pandas/core/algorithms.py

```diff
@@ -698,16 +699,21 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
             result = Series(values)._values.value_counts(dropna=dropna)
             result.name = name
             counts = result.values
+            uniq = Series(values)._values.unique()
 
         else:
             keys, counts = _value_counts_arraylike(values, dropna)
+            uniq = unique(values)
 
             if not isinstance(keys, Index):
                 keys = Index(keys)
             result = Series(counts, index=keys, name=name)
 
     if sort:
         result = result.sort_values(ascending=ascending)
+    elif bins is None:
+        if not isinstance(result.index, ABCCategoricalIndex):
```

Comment:
this should not be necessary

Reply:
It unfortunately is for cases, where the categories contain more elements than the values: https://github.com/pandas-dev/pandas/pull/24302#discussion_r243770038

Reply:
it is not necessary otherwise the impl is incorrect

categoricals uniq already handles this

Reply:
It doesn't seem so as `unique` does not contain `d` in this example:
```

In [31]: s = pd.Series(pd.Categorical(list('baabc'), categories=list('abcd')))                                                                                                                                                                

In [32]: s.value_counts()                                                                                                                                                                                                                     
Out[32]: 
b    2
a    2
c    1
d    0
dtype: int64

In [33]: pd.unique(s)                                                                                                                                                                                                                         
Out[33]: 
[b, a, c]
Categories (3, object): [b, a, c]
```

Reply:
https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/categorical.py#L2267 also mentions that unused categories are not returned.

It would be consistent to not return it in the `value_counts` in the above example as well, but would change the current behaviour... What do you think?